smpmgr/terminal.py

```diff
@@ -0,0 +1,74 @@
+import asyncio
+import logging
+from typing import Final, cast
+
+import readchar
+import typer
+from serial import Serial
+
+from smpmgr.common import Options
+
+logger = logging.getLogger(__name__)
+
+MAP_KEY_TO_BYTES: Final = {
+    readchar.key.CTRL_C: b"\x03",
+    readchar.key.UP: b"\x1b[A",
+    readchar.key.DOWN: b"\x1b[B",
+    readchar.key.LEFT: b"\x1b[D",
+    readchar.key.RIGHT: b"\x1b[C",
+    readchar.key.ESC: b"\x1b",
+}
+
+
+def terminal(ctx: typer.Context) -> None:
+    """Open a terminal to the device."""
+
+    options = cast(Options, ctx.obj)
+
+    async def f() -> None:
+        if options.transport.port is None:
+            print("--port <port> option is required for the terminal, e.g.")
+            print("smpmgr --port COM1 terminal")
+            return
+
+        print(f"\x1b[2mOpening terminal to {options.transport.port}...", end="")
+
+        with Serial(port=options.transport.port, baudrate=115200, timeout=options.timeout) as s:
+            print("OK")
+            print("Press Ctrl-T to exit the terminal.\x1b[22m")
+            print()
+            device_result, keyboard_result = await asyncio.wait(
+                (
+                    asyncio.create_task(_rx_from_device(s)),
+                    asyncio.create_task(asyncio.to_thread(_tx_keyboard_to_device, s)),
```

Comment:
Interesting usage here. Why does the _tx_keyboard_to_device need the extra conversion to a thread?

Reply:
Unfortunately Python has no async means of receiving keyboard events, so I put it on a python thread.  The same is true for pyserial but it’s easier to hack to asyncio since the OS stores a buffer of serial input.

In the case of serial, I know that Windows and Linux can be async, but pyserial does not implement it.

I’m not so sure about keyboard events.  My brief look at Windows APIs didn’t make it obvious that a user can register a callback or event on keyboard events, which AFAIK would be required to make a proper async input module.